api/errors.go

```diff
@@ -0,0 +1,86 @@
+package api
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Application error codes.
+//
+// These are meant to be generic and they map well to HTTP error codes.
+const (
+	ECONFLICT       = "conflict"
+	EFORBIDDEN      = "forbidden"
+	EINTERNAL       = "internal"
+	EINVALID        = "invalid"
+	ENOTFOUND       = "not_found"
+	ENOTIMPLEMENTED = "not_implemented"
+	EUNAUTHORIZED   = "unauthorized"
+)
+
+// Error represents an application-specific error.
+type Error struct {
+	// Machine-readable error code.
+	Code string
+
+	// Human-readable error message.
+	Message string
+
+	// DebugInfo contains low-level internal error details that should only be logged.
+	// End-users should never see this.
+	DebugInfo string
+}
+
+// Error implements the error interface. Not used by the application otherwise.
+func (e *Error) Error() string {
+	return fmt.Sprintf("error: code=%s message=%s", e.Code, e.Message)
+}
+
+// WithDebugInfo wraps an application error with a debug message.
+func (e *Error) WithDebugInfo(msg string, args ...any) *Error {
+	e.DebugInfo = fmt.Sprintf(msg, args...)
+	return e
+}
+
+// ErrorCode unwraps an application error and returns its code.
+// Non-application errors always return EINTERNAL.
+func ErrorCode(err error) string {
```

Comment:
This should probably be a WithError

Reply:
`WithError` would make sense if we were attaching the error. But here we're just extracting the code from the error.